[1단계 - 점심 뭐 먹지] - 클레어(안은소) 미션 제출합니다.

eunsoA opened the pull request:
안녕하세요 마르코! 😃 클레어입니다!
이번 리뷰 잘 부탁드립니다!

## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://eunsoa.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->


1. 컴포넌트 기준
이번 미션은 컴포넌트화 하는 것이 중점이었기 때문에, 컴포넌트에 대해 개인적으로 정의해봤습니다.
  (1) 재사용성 : 두군데 이상에서 사용한다.
  (2) 관심사 분리 : 역할과 기능에 맞게 하나의 기능만 하도록 분리함.
  (3) 응집도 : 유지 보수시 기능 변경이 되어도 다른 기능에 영향을 미치지 않는다.
위의 기준을 기준으로 "컴포넌트는 재사용성이 있어야 한다."라는 것을 초점을 잡고, 프로그램에 필요한 컴포넌트를 모두 나열해보았습니다.
  > 점심 뭐먹지 컴포넌트
  > - 헤더
  > - 레스토랑 아이템
  >    - 레스토랑 아이콘
  >    - 레스토랑 설명
  >- 모달창
  >    - 모달창 헤더
  >    - input - dropdown
  >    - input - textbox
  >        - title
  >        - select
  >        - input
  >        - textarea
  >        - description
  >    - button

2. 에러문 'Argument 1 ('node') to Node.appendChild must be an instance of Node'
이번 미션을 하면서 페어와 가장 많이 만난 에러문입니다. 템플릿을 사용하면서 string이기 때문에 해당 에러를 많이 접했던 것 같습니다.
 

### 2) 이번 리뷰를 통해 논의하고 싶은 부분
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

1. createElemnet, querySelelctor 자제하고, 템플릿 사용
DOM 을 건드려서 생성, 삭제, 쿼리를 들고오는 작업들은 비싼 작업이라고 알고 있습니다. 그래서 최대한 자제하려고 했습니다. 처음에는 innerHTML도 사용하지 않고, documnetFragment나 template 태그를 통해서 구현하려고 했지만, 오히려 불필요한 로직이 더 많이 들어간다고 판단했습니다. 그래서 최대한 innerHTML을 사용하며, DOM을 직접 건드리지는 않으려고 했습니다. 그럼에도 불구하고 불가피하게 몇몇 부분들은 요소를 생성하고, 쿼리를 들고 오고 있습니다. 이 부분에 대해서 마르코의 생각이 궁금해요.
  > 제가 개인적으로 생각한 템플릿에 대한 성능적 이점
  > - createElement와 같은 DOM 조작은 매번 새로운 요소를 생성하고 메모리를 할당해야 하므로 비용이 많이 듭니다.
  > - 템플릿은 HTML 문자열을 한 번에 파싱하므로 더 효율적입니다.
  > - HTML 구조를 직관적으로 볼 수 있어 마크업의 구조 파악이 쉽습니다.

2. 도메인 로직 관련
개발 순서로 따지면 도메인 구현, 정적인 템플릿 - 동적으로 변경 - 스타일 모듈화 - 기능 붙이기 - 리팩터링 정도 순서로 가는 것이 이상적이라고 생각해요. 그런데 이번 미션을 하면서는 ui를 컴포넌트로 분리하는 것에 집중하다보니 우선은 ui를 분리하고, 그다음에 도메인 영역을 후에 작업한 것 같아요. 
그런데 이번 미션의 경우, 이전 로또미션이나 자동차 미션과 달리 도메인 영역이 애매해서 그랬던 것도 같아요. 저는 이번 미션에서 도메인 로직은 추가하는 부분만 이었다고 생각하는데, 마르코가 생각하시기에 이번 미션에서 도메인 로직의 영역이 어디까지로 생각하시는지 궁금합니다!

3. 도메인 로직과 UI로직 분리
위와 이어지는 질문으로 도메인에 대한 영역이 명확하지 않다보니 코드에서도 도메인과 UI가 분명하게 분리되지 않는 것 같아요. 지금 코드에서 도메인과 UI를 더 구분하는 게 맞는건지 궁금해요? 개인적으로 지금은 도메인이 잘 분리 되지 않았다는 생각도 드는데 마르코의 의견이 궁금합니다.

4. E2E 테스트
이번 기회에 cypress를 처음 써봤는데요, 이전에 TDD 기억이 남아있어서 그런지, 테스트 코드를 짜고 보니 단위 테스트와 비슷하다는 느낌이 들어요. 다른 크루와 이야기 해보니 '모달 열기, 모달 닫기, 가게 추가하기' 이렇게 3가지에 대해서만 테스트를 작성한 팀도 있더라구요. 
그렇다면 현재 저희가 짠 테스트 코드는 E2E보다는 단위 테스트에 가까운 걸까요?

5. 2단계 방향성
2단계에서는 아래와 같은 내용으로 수정및  보완하고 방향을 잡았는데요, 여기에서 더 추가 수정할 내용이 있을지 궁금합니다.
  - 2단계 미션 기본 기능 구현
  - localStorage
  - 도메인 로직 분리 (?)
  - 스타일 모듈화에 대해 고민해볼 예정
  - input validation (현재는 추가하기 버튼 클릭시에만 유효성 검증을 하지만, focus가 잡힐때 유효성 검사를 하는 것이 더 사용자 관점에서 친절하다는 생각이 들어서 추가 보완할 예정입니다.)

이것저것 궁금한게 많아서 PR이 길어졌네요..😅
잘부탁드립니다!
---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [x] UI를 컴포넌트 단위로 분리했는가?
- [x] 재사용 가능한 컴포넌트를 고려했는가?
- [x] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

H0ngJu committed:
docs: 구현할 기능 목록 작성

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: eslint, prettier 세팅

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: 사진, css 파일 이동

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: index 초기 템플릿 적용

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat: header 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 헤더 구현 완료

H0ngJu committed:
style: 사용하지 않는 헤더 삭제

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat: 이미지 alt, src 객체 생성

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat: 가게 이름, 거리, 설명, 아이콘 컴포넌트 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 가게 이름, 거리, 설명, 아이콘 컴포넌트 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat: restuarant 객체 생성 및 restuarantList 선언

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat: restaurantItem 컴포넌트 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat: restaurantList 출력

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
fix: CATEGORY export error 수정

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
style: 사용하지 않는 restaurant-list 삭제

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 가게 리스트 출력, 가게 목록 아이템 컴포넌트 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: eslint, prettier 설정으로 인한 e2e 수정

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
refactor(header): header class로 변경

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
refactor(restaurantInfo): restaurant 정보 컴포넌트 클래스화

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
refactor(restaurantIcon): restaurant 아이콘 컴포넌트 클래스화

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
refactor(restaurantItem): restaurant 아이템 컴포넌트 클래스화

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: eslint 브라우저 추가

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(Button): button 컴포넌트 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(modal): modal 컴포넌트 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: console.log 디버깅 삭제 및 컴포넌트 분리로 인한 index.html 수정

H0ngJu committed:
feat(button): converStringToElment 함수 분리

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(AddRestaurantModal): Modal을 상속받아 AddRestaurantModal 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(header): addButton 클릭시 모달창 호출

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 헤더 클릭 이벤트, 모달창 버튼 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(inputDropDown): 모달창에 카테고리, 거리 드롭다운 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 카테고리, 거리 드롭다운 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(inputText): 모달창에 이름, 설명, 참조 링크 input 컴포넌트 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(AddRestaurantModal): modalTitle 추가

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 모달창 입력, 설명, 링크, 입력 아래 설명, 서브 타이틀 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(AddRestaurantModal): 취소하기 버튼 클릭시 모달 창 닫는 기능 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 취소하기 버튼 클릭시 모달창 닫기 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
fix: 모달 form 태그 추가

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(addRestaurantModal): 추가하기 버튼 클릭 시 리스트에 새 음식점 추가 기능 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
style: distance 출력 포맷 수정

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 새로운 식당 리스트에 추가 기능 구현 완료

H0ngJu committed:
refactor(addRestaurantModal): 가독성을 위한 init(), appendChildToModalForm(), addHandler(), removeModal() 함수 분리

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
feat(validation): 카테고리, 이름, 거리, 설명, 링크에 대한 유효성 검증 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
style: 드롭다운 디폴트 메시지 수정

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 카테고리, 이름, 거리, 설명, 링크에 대한 유효성 검증 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: header -> Header 파일명 변경

H0ngJu committed:
chore: restaurantItem -> RestaurantItem 파일명 변경

H0ngJu committed:
feat(AddRestaurantModal): 백그라운드, esc 누름 시 모달창 닫기 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 백그라운드, esc 누름 시 모달창 닫기 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
refactor(inputDropDown): value 넘겨주도록 수정

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
test: 음식점 목록 페이지, 모달창 선택 테스트 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
fix(inputDropDown): 거리 value, innerValue 순서 변경

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
design: 모달창 css 수정

H0ngJu committed:
refactor(Modal): modal--open 토글 방식 변경

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: 불필요한 주석 및 log 삭제

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: 불필요한 주석 및 log 삭제

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
test: e2e 테스트를 위한 상수 선언

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
test: 초기 화면 렌더링 테스트

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
test: 헤더 이벤트 테스트

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
test: 모달 동작 관련 테스트

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
test: 모달 입력 관련 테스트

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: E2E 테스트 목록 업데이트

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
style: 거리 텍스트 수정

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
test: 새로고침 테스트 추가

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 새로고침 테스트 구현 완료

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: 배포 링크 수정

H0ngJu committed:
chore: 배포 링크 소문자로 변경

H0ngJu committed:
chore: 배포 이름 변경

H0ngJu committed:
chore: html 경로 수정

H0ngJu committed:
chore: import 문 수정

H0ngJu committed:
chore: img 링크 수정

H0ngJu committed:
chore: img 링크 수정

H0ngJu committed:
test: 모달 alert 테스트 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 모달 alert 테스트 구현

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
docs: 프로젝트 구조 업데이트

Co-authored-by: Claire <eunsoA@users.noreply.github.com>

H0ngJu committed:
chore: img 경로 수정

eunsoA committed:
chore: 이미지 경로 상대 경로로 수정

eunsoA committed:
chore: 배포 링크 수정

wonsss commented on README.md:
```diff
@@ -0,0 +1,216 @@
+  - 거리는 캠퍼스로부터 도보로 걸리는 시간(분). 5, 10, 15, 20, 30 중 하나를 선택한다.
+  - 설명, 참고 링크는 옵션. 입력하지 않아도 음식점을 추가할 수 있어야 한다.
+  - 입력값이 잘못되었을 때 사용자에게 알려주는 방식은 자유롭게 구현한다.
+- 새로고침 시 이전에 추가한 새로운 음식점 정보는 초기화된다.
```
현재 form submit 후 이전 form data가 reset 되지 않고 있는데, 일반적인 사용성이 아니라서 이 부분도 개선해보시면 좋을 것 같아요.

https://github.com/user-attachments/assets/87e56e51-060c-4578-9730-cb0132225d77

eunsoA replied:
reset 하는 부분이 빠졌군요! 확인해서 수정했습니다! 감사합니다!
추가로, 모달창이  열리지 않은 상태에서 esc키를 누르면 모달창이 열리는 문제도 있어서 수정하여 반영했습니다!

(commit : [84e64d5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/84e64d5858001e7634a004416eaaaaa3e6a20e69))

wonsss commented on src/components/InputText.js:
```diff
@@ -0,0 +1,54 @@
+    formItem.classList.add('form-item');
+    formItem.innerHTML = `${LABEL_TEMPLATE('description', '설명')}${TEXTAREA_TEMPLATE()}${SPAN_TEMPLATE(
+      '메뉴 등 추가 정보를 입력해 주세요.',
+    )}`;
```
innerHTML은 XSS 공격을 받을 수 있어요.
한번 현재 코드에서 설명 인풋에 `<img src="x" onerror="alert('XSS 공격!')" /> ` 를 입력해서 form 제출하시면, 해당 자바스크립트 코드가 실행되어 alert 가 뜨는 것을 보실 수 있어요.
따라서 외부에서 입력받는 신뢰할 수 없는 데이터는 sanitization 시키거나 textContent 를 사용해야 해요.

eunsoA replied:
가독성의 문제만 생각했는데 innerHTML 의 보안 취약점에 대한 내용은 인지하지 못했었네요..! 아래에 수정된 내용 작성해놨어요. 감사합니다!

wonsss commented on src/components/InputDropDown.js:
```diff
@@ -0,0 +1,43 @@
+      <option value="">선택해주세요</option>
+    </select>
+  `;
+};
```
> createElemnet, querySelelctor 자제하고, 템플릿 사용
DOM 을 건드려서 생성, 삭제, 쿼리를 들고오는 작업들은 비싼 작업이라고 알고 있습니다. 그래서 최대한 자제하려고 했습니다. 처음에는 innerHTML도 사용하지 않고, documnetFragment나 template 태그를 통해서 구현하려고 했지만, 오히려 불필요한 로직이 더 많이 들어간다고 판단했습니다. 그래서 최대한 innerHTML을 사용하며, DOM을 직접 건드리지는 않으려고 했습니다. 그럼에도 불구하고 불가피하게 몇몇 부분들은 요소를 생성하고, 쿼리를 들고 오고 있습니다. 이 부분에 대해서 마르코의 생각이 궁금해요.

innerHTML을 사용하는 게 성능상 더 비싼 작업이에요.(벤치마크 참고) 또한 보안 이슈도 있구요. 그래서 이러한 문제의식이라면 createElement와 appendChild를 사용하여 요소를 구성하는 게 적절하다고 생각해요.

다만, 이렇게 했을 때, 가독성이 문제라면 템플릿 리터럴을 사용하는 방향으로 다시 고민해보실 수 있는데,  그렇다면 가독성과 성능상 이점이 있는 template 태그를 활용하여 DOM을 효율적으로 조작하는 방식을 고려해보는 것도 좋아요. template 태그로 구현하시려했으나 불필요한 로직이 많이 들어간다고 느끼셨다고 했는데, innerHTML은 아주 단순한 반면 위에서 언급한 단점들의 트레이드오프가 있는 것 같아요.

> 참고 https://javascript.plainenglish.io/stop-using-innerhtml-heres-the-secret-technique-netflix-engineers-use-for-dom-manipulation-5a53df44ba9e

eunsoA replied:
저는 템플릿 형식으로 코드를 구현했다고 생각했는데, 말씀해주신 내용과 자료들을 참고해서 보니 제가 사용한 방식은 템플릿 태그를 사용한 방식도 아니더라구요 ..! 허헛 .. 

그래서 템플릿 태그를 사용하는 방식을 다시 알아보고 템플릿 태그 방식과 createElemetn와 appendChild를 사용하는 방식의 장단점을 비교해봤습니다. 

- 템플릿 태그와 createElement와 appendChild 방식 모두 재사용은 가능함. 다만 템플릿의 경우, HTML 파일에 의존하기 때문에(html에 template태그를 넣어놓고 재사용하는 방식임.) 독립적인 컴포넌트가 목적이라면 createElement와 appendChild 방식이 더 적합함.
- 또한 동적으로 컴포넌트를 생성하고 수정한다는 관점에서 컴포넌트라는 목적에는  createElement와 appendChild 방식이 더 적합하다고 판단함.
- 마지막으로 가장 큰 장점:  createElement와 appendChild 방식은 컴포넌트를 독립적으로 테스트할 수 있음.

그래서 내린 제 결론은 아래와 같아요!
- 템플릿 태그 대신에 createElement와 appendChild 방식을 사용한다.

그리하여 수정한 커밋은 아래와 같습니다. 기존의 innerHTML을 사용한 코드는 아래와 같이 수정했어요.
* Header (commit - [cba1582](https://github.com/woowacourse/javascript-lunch/pull/209/commits/cba15828cd0456d0a4881884abe347b2527b6fc8))
* Button (commit - [4df3ce5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/4df3ce52e41cefed4fbbe3152c0742c09768e528))
* InputDropDown, InputText, Modal(commit - [84e64d5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/84e64d5858001e7634a004416eaaaaa3e6a20e69))
* RestaurantIcon (commit - [90d88b8](https://github.com/woowacourse/javascript-lunch/pull/209/commits/90d88b88333d8b0f3190286e301bb9e532e69544))
* RestaurantInfo (commit - [23c902a](https://github.com/woowacourse/javascript-lunch/pull/209/commits/23c902afafa365b95900df3e8d2be98be8c0403c))

마르코가 말씀해주신 덕분에 템플릿 태그 방식과 createElement와 appendChild 방식 의 차이를 명확히 이해헀어요. 감사합니다!

eunsoA replied:
추가로 현재 modal을 상속받는 addRestaurantModal는 super의 메소드를 사용함으로 응집도가 떨어진다는 생각이 들어요. 마르코가 보기엔 어떠신가요?  이걸 좀더 각자의 역할이 응집될 수 있도록 수정할 방법이 있을지 궁금해요.

wonsss replied:
정리를 정말 잘하시네요! 👍 훌륭합니닷

wonsss commented on src/modal/AddRestaurantModal.js:
```diff
@@ -0,0 +1,134 @@
+    const newRestaurant = new Restaurant(testData.name, testData.distance, testData.description, testData.category);
+    const newRestaurantItem = new RestaurantItem(newRestaurant);
+    DOM.RESTAURANT_LIST.appendChild(newRestaurantItem);
+  };
```
> 도메인 로직과 UI로직 분리
위와 이어지는 질문으로 도메인에 대한 영역이 명확하지 않다보니 코드에서도 도메인과 UI가 분명하게 분리되지 않는 것 같아요. 지금 코드에서 도메인과 UI를 더 구분하는 게 맞는건지 궁금해요? 개인적으로 지금은 도메인이 잘 분리 되지 않았다는 생각도 드는데 마르코의 의견이 궁금합니다.

AddRestaurantModal 은 현재 UI 로직과 도메인 로직이 섞여 있는 구조네요.
AddRestaurantModal은 UI를 담당하도록 하고, 이와 같이 '음식점 추가'하는 도메인 로직은 별도의 모듈로 분리하는 게 가독성이나 유지보수에 용이합니다. 음식점 데이터를 관리하고 '음식점 추가', '수정', ' 삭제' 등의 역할을 할 수 있는 RestaurantService 같은 모듈을 분리하는 게 더 나은 설계라고 볼 수 있어요.
UI 컴포넌트에서 도메인 로직을 호출하고, 도메인 로직에서 음식점 추가 후 새로운 음식점 목록 데이터를 반환하면, View에서는 그  데이터를 받아서 화면에 그려주기만 하면 되도록요. 
이전 미션에서 학습했듯이, View는 변경되더라도 도메인 로직은 유지되는 구조가 안정적이므로 결합을 분리해보면 좋을 거 같습니다.

eunsoA replied:
도메인 로직 분리에 대해 생각하지 못하고 UI 컴포넌트 분리에만 집중했던 것 같아요. 수정해서 반영했습니다! (commit :[656865e](https://github.com/woowacourse/javascript-lunch/pull/209/commits/656865e85aa6023119fe7f8cdebd289df3fcc912),  [88ee8aa](https://github.com/woowacourse/javascript-lunch/pull/209/commits/88ee8aaa15f1ba286ec768dddedae54d1a5264ca))

추가로 현재 도메인 로직은 RestaurantService 에서 함수형으로 구현했는데요, 마르코가 보기에 수정해야할만한 사항이 있을까요?

wonsss replied:
클레어, 수정 잘해주셨네요. 👍  함수형으로 변경한 부분도 좋고, 전반적으로 코드 구조가 더 깔끔해졌어요. 현재 코드처럼 RestaurantList.js와 RestaurantService.js로 나누어도 동작에는 문제가 없지만, 데이터와 로직을 좀 더 명확히 분리하는 것이 좋을 것 같아요.

지금 RestaurantList.js에서 리스트를 직접 export하면서 싱글톤처럼 동작하고 있는데, 이 파일은 단순히 데이터를 보관하는 역할만 하고 있어요. 그런데 대부분의 코드는 mock 데이터를 정의하는 데 사용되고 있어서, 이런 데이터는 별도의 파일로 분리하는 것이 더 명확한 구조가 될 것 같아요. 예를 들면 MockRestaurantData.js 같은 파일을 만들어 관리하면 더 깔끔할 거예요.

또한, 응집성을 높이기 위해 RestaurantList.js 파일을 제거하고, RestaurantService 내부에서 restaurants 상태를 관리하는 것이 더 좋을 것 같아요. 이렇게 하면 데이터 관리와 로직이 한 곳에서 이루어지므로 코드의 가독성이 좋아지고 유지보수도 쉬워질 거예요.  특히, RestaurantService를 클래스로 만들어 restaurants를 클래스 필드로 두어 restaurants 배열을 직접 노출하지 않고 상태를 내부에서 관리함으로써 데이터 변경을 더 명확히 추적할 수도 있을 것 같아요.

참고만 해주셔도 되고, 다음 미션 시 반영해주셔도 될 것 같아요. 😄

eunsoA replied:
오 클래스의 필드로 두고 관리하는 게 더 나을 수 있겠네요! 이건 고민해보고 2단계에서 반영할게요. 감사해요 :)

wonsss commented on src/modal/AddRestaurantModal.js:
```diff
@@ -0,0 +1,134 @@
+    } catch (error) {
+      alert(error.message);
+    }
+  };
```
유효성 검사 또한 도메인 로직으로 분리하는 게 좋아요

eunsoA replied:
넵 요것또한 동의해요!
수정해서 반영했습니다!  (commit :[656865e](https://github.com/woowacourse/javascript-lunch/pull/209/commits/656865e85aa6023119fe7f8cdebd289df3fcc912))

wonsss commented on src/modal/AddRestaurantModal.js:
```diff
@@ -0,0 +1,134 @@
+  };
+
+  #addHandler = () => {
+    const testData = Object.fromEntries(new FormData(this.#modalForm));
```
testData라는 네이밍을 개선해볼 수 있을 것 같아요.
testData라고 하면 일반적으로 그냥 테스트용 데이터를 의미하는 경우가 많아서 오해를 불러일으킬 수 있어요.
그리고 이게 무슨 의도의 변수인지 단순히 'test', 'data'만 봐서는 알 수가 없어서, 명확하게 수정하면 좋겠어요.

eunsoA replied:
으악 페어프로그래밍때 일단 테스트로 적어보자~하고 했던 네이밍이 변경이 안됐네요! 말씀해주셔서 감사합니다! 확인하고 수정했어요! (commit : [22e3304](https://github.com/woowacourse/javascript-lunch/pull/209/commits/22e3304040a5e334ce0dcdb009a87a3c9f7ef337))

wonsss replied:
변수 네이밍이 훨씬 개선됐네요 👍

wonsss commented on src/modal/AddRestaurantModal.js:
```diff
@@ -0,0 +1,134 @@
+  };
+
+  #validateInputs = () => {
+    const testData = Object.fromEntries(new FormData(this.#modalForm));
```
마찬가지로 전체적으로 testData나 모호한 변수 네이밍 개선 하면 좋겠어요

eunsoA replied:
얘도요! (commit : [22e3304](https://github.com/woowacourse/javascript-lunch/pull/209/commits/22e3304040a5e334ce0dcdb009a87a3c9f7ef337))

wonsss commented on src/modal/AddRestaurantModal.js:
```diff
@@ -0,0 +1,134 @@
+    this.#appendChildToModalForm();
+    this.addElement(this.#modalForm);
+
+    const divButton = this.#createButton();
```
createButton 이라는 메서드의 이름이나 divButton이라는 요소의 네이밍도 의도를 명확하게 더 담아도록 수정할 수 있을 것 같아요.
버튼을 한 개만 만드는 게 아니라, 버튼 그룹을 생성하는 메서드이므로 단순하게는 createButtonGroup 과 buttonGroup 같은 네이밍이 적절할 수 있어요.

eunsoA replied:
페어와 논의하면서 시간이 부족해서 네이밍에 신경을 많이 못썼던 것 같아요 
마르코의 말대로 div라는 접두사는 설명이 부족하다는 것에 너무나도 동의해요! 저또한 이게 무슨 메소드인지 무슨 변수인지 다시 읽어보게 만드니까요.

아래와 같이 수정했어요! (commit : [84e64d5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/84e64d5858001e7634a004416eaaaaa3e6a20e69))
* divButon -> buttonContainer
* createButton -> createButtonContainer

wonsss commented on src/modal/AddRestaurantModal.js:
```diff
@@ -0,0 +1,134 @@
+
+    document.querySelector('.modal-backdrop').addEventListener('click', () => {
+      this.closeModal();
+    });
```
이벤트 위임을 적용해볼 수 있겠네요~!

eunsoA replied:
오우 이벤트 위임을 적용하면 2단계 미션에 있을 모달까지 생각했을때, 중복성도 적고 유지보수하기에도 더 좋을것같네요! 
반영해서 수정하였습니다! (commit : [95d81f4](https://github.com/woowacourse/javascript-lunch/pull/209/commits/95d81f44af6d6f17ddecff08909967ed72de3bd9))

eunsoA replied:
추가로, escape keyup에 대해서도 이벤트 위임을 적용할 수 있을것 같아서 수정했습니다! (commit : [5857a34](https://github.com/woowacourse/javascript-lunch/pull/209/commits/5857a34b2a50299e8bf7d8629ebc8ee971a1e021))

eunsoA replied:
추가로 현재 modal을 상속받는 addRestaurantModal는 super의 메소드를 사용함으로 응집도가 떨어진다는 생각이 들어요. 마르코가 보기엔 어떠신가요? 이걸 좀더 각자의 역할이 응집될 수 있도록 수정할 방법이 있을지 궁금해요.

wonsss replied:
addRestaurantModal 같은 상속(inheritance) 방식보다는 조합(composition) 방식을 사용하는 게 더 유연하고 재사용성이 높아져요. 

> https://tecoble.techcourse.co.kr/post/2020-05-18-inheritance-vs-composition/

wonsss commented on src/components/Header.js:
```diff
@@ -0,0 +1,31 @@
+`;
+
+class Header {
+  constructor(addRestaurantModal) {
```
Header 컴포넌트에 addRestaurantModal 자체를 주입해주면 불필요하게 큰 인터페이스를 넘기는 거라서, 재사용성을 위해 적절하게 좁혀보면 좋겠어요.
헤더는 addRestaurantModal을 전달받아서, 사실 내부 버튼 클릭 시 addRestaurantModal.openModal() 메서드가 호출되도록 이벤트 핸들러를 등록해두었을 뿐이죠.
그러면 매개변수 이름을 onButtonClick 같이 수정하여, Header 자체는 그냥 버튼 클릭 핸들러를 전달받는다는 사실만 알고 이게 음식점모달을 여는 사실까지는 모르도록 하는 게 재사용에 유리해요.

아니면 핸들러 로직이 부착된 Button 컴포넌트 자체를 전달받아 조합(composition)하는 형식으로 해도 좋구요!

eunsoA replied:
맞아요 이부분도 이것보다는 나은게 있을텐데 라고 생각했는데 두가지 방법이 있네요! 아래와 같이 두가지 방법을 비교하고 결론내린 결과, 이벤트 핸들러 함수를 주입하는 방식으로 결정했어요.
 
 1. 이벤트 핸들러 함수를 주입받는 방식
    1. 더 간단하고 명확한 인터페이스를 제공함
    2. 단일 책임 원칙에 더 부합함
    3. 헤더 컴포넌트의 재사용성이 더 높아짐
2. 버튼 컴포넌트 자체를 주입받는 방식
 
 수정하여 반영했습니다! (commit : [90852fb](https://github.com/woowacourse/javascript-lunch/pull/209/commits/90852fb008b6a87d580b1d31eeac12869f4769d8))

wonsss replied:
잘 수정해주셨네요

wonsss commented on src/components/Header.js:
```diff
@@ -0,0 +1,31 @@
+  <button type="button" class="gnb__button" aria-label="음식점 추가">
+    <img src="./public/assets/add-button.png" alt="음식점 추가">
+  </button>
+`;
```
헤더를 다른 페이지에서도 동일한 UI로 재사용해야 한다면, 어떻게 컴포넌트를 구성하는 게 좋을까요?
현재는 Header 컴포넌트는 '점심 뭐 먹지' 페이지에서만 사용할 수 있어요. 
이처럼 다른 페이지에서도 동일한 UI의 헤더를 재사용해야 하는 요구사항이 있다면, 인터페이스로 title과 button을 전달받도록 하고 Header는 UI 배치 역할을 하는 컴포넌트로 리팩터링하는 것도 고민해볼 수 있을 것 같아요.

eunsoA replied:
오 "이번 미션에서는 사실 헤더가 하나의 스타일로 고정되긴 하지만, 컴포넌트로 만들어서 추후 유지보수하는데에 편리하도록 하자"가 목표였는데요. 지금의 구조는 하나의 컴포넌트라기보다는 점심뭐먹지 미션에서만 쓸 수 있는 구조네요..

만약 다른 형태의 헤더가 있어서 두가지 스타일의 헤더를 구현해야한다면 마르코의 말대로 인터페이스로 title과 button을 전달받아서 헤더는 UI배치 역할만 하는 컴포넌트로 리팩터링하는 것에 동의해요. 
요건 이번 미션에서 당장 급한 수정사안은 아니라는 생각이 들어서.. 우선 급한 내용들 먼저 수정해볼게요!

eunsoA replied:
라 고했지만 위의 리뷰를 반영하면서 고쳐봤어요. (commit : [b7e18dc](https://github.com/woowacourse/javascript-lunch/pull/209/commits/b7e18dc165f05d4f67657b58c48c31eaecef9586))

RestaurantItem처럼 따로 RestaurantIcon와 RestaurantInfo를 객체로 생성하고 appendChild할 필요는 없겠다는 생각이 들어서 우선은 위와 같이 만들었어요. 
이번 미션에서는 헤더가 변경되는 일은 없지만, 만약 헤더가 수정되어야하는 요구사항이 있다면, Header의 createHeader 메서드에서 헤더의 종류에 따라 다르게 createTitle을 하거나, createButton을 해주면 되지 않을까! 라고 생각했는데요,

이렇게 글을 쓰면서 생각해보니 일반적인 헤더의 경우, 
1. 타이틀만 있는경우
2. 타이틀과 addButton이 있는경우
3. 타이틀과 다른Button이 있는경우

이렇게 나눌수있을것같아요. 그렇다면 Header는 인터페이스로 타이틀또는  타이틀과 버튼을 받아서 배치하는게 더 유연할것같아요

마르코의 생각은 어떤가요?

wonsss replied:
잘 수정해주신 것 같아요. 요구사항은 상황마다 다를 수 있어서, 매번 요구사항과 확장 가능성을 잘 고민하고 컴포넌트를 설계하는 게 중요한 거 같아요!

eunsoA replied:
네 이해했습니다!

wonsss commented on src/modal/AddRestaurantModal.js:
```diff
@@ -0,0 +1,134 @@
+    this.#divCategory = new InputDropDown('카테고리', CATEGORY_LIST);
+    this.#divName = new InputText('이름');
+    this.#divDistance = new InputDropDown('거리(도보 이동 시간)', DISTANCE_LIST);
+    this.#divDescription = new InputText('설명');
```
요소들의 역할을 명확하게 나타낼 수 있도록 네이밍 개선할 수 있어요.
div라는 접두사는 의미가 없어서 네이밍에 div를 사용하지 마시고 대상과 역할을 명확히 드러내는 네이밍을 하시는 게 좋아요. 실제로 Dropdown, Input 등의 역할을 지녔는데 역할에 맞지 않는 div라는 네이밍이 가독성을 떨어트리고 있어요.
예를 들어, divCategory -> categoryDropdown, divName -> nameInput 처럼 개선해보시면 좋겠어요.

eunsoA replied:
으악 이 부분도 수정했습니다! (commit : [84e64d5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/84e64d5858001e7634a004416eaaaaa3e6a20e69))

* divCategory  -> categoryDropDown
* divName -> nameInput
* divDistance -> distanceDropDown
* divDescription -> descriptionInput
* divLink -> linkInput

wonsss commented on src/modal/AddRestaurantModal.js:
```diff
@@ -0,0 +1,134 @@
+    this.#init();
+    this.#bindEvent();
+    this.#createAddModal();
+    return this;
```
생성자 메서드에서 this 반환하신 이유가 궁금해요

eunsoA replied:
이것 또한 페어프로그래밍을 하며 구현에 집중하느라 미처 못 수정한 것 같아요.
클래스의 생성자는 자동으로 새로 생성된 인스턴스(this)를 반환하기에 명시적으로 return this를 작성하지 않아도 동일하게 동작하는걸로 알고 있기에 위의 코드에서 this를 반환할 필요는 없어보여요.
확인하고 수정했습니다!
(commit : [84e64d5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/84e64d5858001e7634a004416eaaaaa3e6a20e69))

wonsss commented on src/components/Button.js:
```diff
@@ -0,0 +1,17 @@
+class Button {
+  constructor(className, buttonText) {
+    return this.#crerateButton(className, buttonText);
+  }
```
Button 인터페이스로 className을 받아서 인자로 스타일을 외부에서 자유롭게 제어하도록 열어두는 방식도 다양한 요구사항을 만족하려면 필요할 수 있어요.
다만, 현재 요구사항에 따르면, Button 이라는 컴포넌트는 UI로는 두 가지 타입(primary, secondary)이 존재하므로 인자로 이를 받아서 내부에서 인자로 받은 각 타입별로 스타일을 적용해주는 것도 적절하다고 볼 수 있어요.

eunsoA replied:
오우 이부분은 너무나도 동의해요! 
innerHTML에서 createElement 방식으로 변경하면서 해당 요구사항도 만족하도록 반영했어요 :) (commit : [4df3ce5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/4df3ce52e41cefed4fbbe3152c0742c09768e528))

wonsss commented on src/components/Button.js:
```diff
@@ -0,0 +1,17 @@
+import { convertStringToElement } from '../utils/convertStringToElement.js';
+
+const BUTTON_TEMPLATE = (className, buttonText) => {
+  return `<button type="button" class="button ${className} text-caption">${buttonText}</button>`;
```
type="button"으로 하드코딩해두셨는데, Button 컴포넌트에서 type 은 사용 시 제어할 필요가 있으므로 인터페이스로 받는 게 좋아요.
현재 form submit 역할의 버튼('추가하기')도 이 Button 컴포넌트를 사용했기 때문에, submit type을 주지 못했고 submit 이벤트로 제어하지 못해 웹접근성이 떨어진 부분이 있어요.

eunsoA replied:
음.. 템플릿을 쓰고싶다는 생각에 템플릿 구조를 구현하는데만 집중했더니 이런 점은 전혀 고려하지 못했네요..!

그런데 한가지 궁금한 점이 있어요!

현재 addRestaurantModal의 Button의 경우, button을 사용하여 취소하기와 추가하기 버튼을 만들고 있어요. 그런데 현재 form submit 역할의 버튼('추가하기')도 이 Button 컴포넌트를 사용했기 때문에, submit type을 주지 못했고 submit 이벤트로 제어하지 못해 웹접근성이 떨어진 부분이 있다는 것은 이해돼요. 그러나 제가 알기로는 submit 태그를 목적에 맞게 쓰려면 폼 내부에 버튼 태그가 있어야 하는걸로 알고 있어요.

그런데 저는 화면에 모달창이 렌더링됐을때, 폼 태그는 스크롤이 되고, 버튼은 모달창이 작아서 아래로 밀리지 않도록, 버튼은 폼 태그의 밖으로 뱄어요. (<- 모달창 내용이 가려지더라도 버튼은 보이는게 사용자 입장에서 더 친화적이라는 생각이 들어서 페어와 논의 후 결정했었습니다.)

그렇게 되면, 폼 내부 태그에 버튼이 있는게 아니라서 submit 태그가 있어도 웹 접근성면에서 의미가 없는 게 되는건 아닌가요? 아런 경우, 어떻게 하는지 궁금해요.

eunsoA replied:
음 .. ! 그런데 방법을 찾았습니다. 
폼내부에 버튼을 넣되, 내부에서 스크롤이 되는 영역(버튼을 제외한 영역) div로 분리해서 css를 추가하였어요.
(commit : [c4c4b9d](https://github.com/woowacourse/javascript-lunch/pull/209/commits/c4c4b9d45fd5e88b8ad11290f1ee7f35549b374a))

wonsss replied:
방법을 찾으셨군요 👍

wonsss commented on src/components/Button.js:
```diff
@@ -0,0 +1,17 @@
+};
+
+class Button {
+  constructor(className, buttonText) {
```
그리고 onClick 인터페이스도 추가해보는 것도 고민해볼 수 있어요. 
현재는 버튼을 생성한 후, 외부에서 addEventListener를 따로 추가해야 하는데, 인터페이스를 추가하면 클래스 내에서 바로 클릭 이벤트를 설정할 수 있어요. 이를 통해 Button을 사용할 때, new Button({text, onClick}) 처럼 한번 이벤트까지 설정할 수 있어 더 직관적인 코드가 됩니닷.

eunsoA replied:
오 인터페이스를 추가해서 버튼생성시 이벤트 리스너에 추가해주는 방식이 더 직관적인것같아요!
반영해서 수정했습니다! (commit : [09b560a](https://github.com/woowacourse/javascript-lunch/pull/209/commits/09b560a6c48f4e4d2be909abc2835ae9bb8608d2))

wonsss replied:
잘 수정해주셨네요. 참고로 listener 라는 네이밍보다는 해당 리스너가 click 이벤트에 부착될 것이므로, onClick 같이 명시적인 역할을 표시할 수 있는 네이밍을 해주는 게 가독성에 좋아요.

eunsoA replied:
오 조언 감사합니다! 2단계에서 반영할게요!

wonsss commented on src/style/style.css:
css도 하나의 style.css 에 모두 모여 있네요.
컴포넌트를 분리하시면서, 각 컴포넌트에 해당하는 css들은 별도의 모듈로 분리하여 개별 컴포넌트 폴더에 응집해두면 유지보수할 때 편리하실 것 같아요.

eunsoA replied:
앗 네 요건 반영해서 수정했습니다! (commit : [a20798b](https://github.com/woowacourse/javascript-lunch/pull/209/commits/a20798b5fe6734de4be9dae36c8907029220d167))

wonsss commented on src/components/Modal.js:
```diff
@@ -0,0 +1,36 @@
+    this.#modal = this.createModal();
+  }
+
+  createModal = () => {
```
private method여야 할 것 같아요

eunsoA replied:
오우 이 부분은 놓쳤던 것 같네요. 수정해서 반영했습니다! 감사해요 
(commit : [84e64d5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/84e64d5858001e7634a004416eaaaaa3e6a20e69))

wonsss commented on src/components/Modal.js:
```diff
@@ -0,0 +1,36 @@
+  };
+
+  addElement = (element) => {
+    const divElement = this.#modal.querySelector('.modal-container');
```
addElement 호출할 때마다 querySelector를 호출하게 될텐데, modalContainer도 클래스 필드로 저장하여 관리하면 조금 더 성능 개선될 것 같아요.

eunsoA replied:
오우 그렇네요! 동의합니다! 
해당 내용도 반영해놨어요. 감사합니다!
(commit : [84e64d5](https://github.com/woowacourse/javascript-lunch/pull/209/commits/84e64d5858001e7634a004416eaaaaa3e6a20e69))

추가로 같은 맥락으로 modalContainer에 대해서도 클래스 필드로 저장하여 관리하는 것으로 수정했습니다! (1단계 PR에서는 클래스 필드로 저장했다가 다시 querySelector를 호출하는 방식으로 수정했다가 .. 다시 클래스 필드로 저장하는 방식으로 돌아왔어요.)
(commit : [ed4e9c4](https://github.com/woowacourse/javascript-lunch/pull/209/commits/ed4e9c41849a1d496e19953f5f5d6fb02fd715ec))

wonsss replied:
좋습니다

wonsss commented on cypress/e2e/modalAlert.cy.js:
```diff
@@ -0,0 +1,115 @@
+import { ERROR_MESSAGES } from '../../src/validation/validations.js';
+import { TEST_CONSTANT } from './constants/testConstant.js';
+
+describe('필수 입력 조건이 만족되지 않았을때, 모달창에서 추가하기 버튼을 클릭하면 Alert창을 띄운다.', () => {
```
> E2E 테스트
이번 기회에 cypress를 처음 써봤는데요, 이전에 TDD 기억이 남아있어서 그런지, 테스트 코드를 짜고 보니 단위 테스트와 비슷하다는 느낌이 들어요. 다른 크루와 이야기 해보니 '모달 열기, 모달 닫기, 가게 추가하기' 이렇게 3가지에 대해서만 테스트를 작성한 팀도 있더라구요.
그렇다면 현재 저희가 짠 테스트 코드는 E2E보다는 단위 테스트에 가까운 걸까요?

현재 작성한 cypress 테스트코드들은 E2E라고 보기엔 약간 애매한 부분이 있어요. 경우에 따라 이처럼 단위 테스처럼 cypress 활용해볼 수도 있어요.

다만, E2E 테스트로 보기 어려운 이유는 사용자의 실제 흐름(End-to-End)가 포함되지 않았기 때문이에요.
예를 들어, "사용자가 음식점 추가 버튼을 클릭하여 열린 모달의 양식에 음식점 정보를 입력한 후, 추가 버튼을 눌렀을 때 정상적으로 목록에 추가되는지" 와 같은 전체적인 흐름을 테스트하는 게 E2E라고 볼 수 있어요.
이번 미션의 목표대로 사용자의 실제 행동을 시뮬레이션하는 시나리오를 추가하여 E2E 테스트를 보강해보면 좋겠어요~!

eunsoA replied:
음 아무래도 현재 테스트코드는 흐름 테스트라기보다는 단위 테스트에 가까운 것 같아요. 아래와 같이 3단계로 나눠서 수정했습니다! (commit : [4963688](https://github.com/woowacourse/javascript-lunch/pull/209/commits/496368875488f16139be1ac4aa02144a88de21ef))
1. 사용자가 음식점 목록 페이지에서 음식점 추가 기능을 사용할 수 있다.
  1. 사용자가 음식점 목록에서 우측 상단의 추가 버튼을 눌러 모달 창을 띄운다.
  2. 사용자가 모달 창을 닫을 수 있다.
  3. 사용자가 모달창의 필수 입력 항목을 입력한 후 추가 버튼을 누르면 음식점 목록에 추가된다.

다만,한가지 궁금한 내용은 현재는 사용자가 올바르게 작성한 경우의 흐름만 테스트하는데, 올바르지 않은 흐름에 대해서도 테스트를 추가해야할까요?

wonsss replied:
넵! 올바르지 않은 흐름에 대해서 예외 처리를 하셨다면, 그러한 예외 처리도 비즈니스 로직이라고 볼 수 있어요. 예상대로 잘 예외 처리가 됐는지 보는 게 사용자 경험에 중요하다고 판단하시고 매번 체크하는 게 번거롭다면 E2E에 추가할 가치가 있다고 생각합니다.
물론, 배보다 배꼽이 더 크지 않게 E2E 테스트는 우선순위를 잘 판단하셔서 중요한 것부터 하시는 게 좋아요.

eunsoA replied:
아 이해했습니다!
현재 e2e 코드에서 (이전에 단위 테스트처럼 만든 것을 제외하고) e2e.cy.js에는 예외처리 로직은 들어가있지않아서 2단계 미션중 여유가 된다면 추가해보겠습니다.

wonsss commented on src/main.js:
> 도메인 로직 관련
개발 순서로 따지면 도메인 구현, 정적인 템플릿 - 동적으로 변경 - 스타일 모듈화 - 기능 붙이기 - 리팩터링 정도 순서로 가는 것이 이상적이라고 생각해요. 그런데 이번 미션을 하면서는 ui를 컴포넌트로 분리하는 것에 집중하다보니 우선은 ui를 분리하고, 그다음에 도메인 영역을 후에 작업한 것 같아요.
그런데 이번 미션의 경우, 이전 로또미션이나 자동차 미션과 달리 도메인 영역이 애매해서 그랬던 것도 같아요. 저는 이번 미션에서 도메인 로직은 추가하는 부분만 이었다고 생각하는데, 마르코가 생각하시기에 이번 미션에서 도메인 로직의 영역이 어디까지로 생각하시는지 궁금합니다!

개발 순서는 사람마다 다를 수 있지만, 도메인 로직과 UI 로직을 분리하는 게 가장 중요한 원칙이에요.
이번 미션에서는 UI 컴포넌트 분리에 집중하다 보니 도메인 로직이 후순위로 다뤄졌고,
특히 이전 로또 미션이나 자동차 미션과 비교했을 때 도메인 영역이 애매하게 느껴진 이유는 UI 중심의 기능이 많기 때문이에요.

그렇다고 도메인 로직이 없는 건 아니에요.
이번 미션에서 도메인 로직을 정리해보면,
음식점 목록 관리, 음식점 추가(중복 검사, 데이터 저장), 음식점 필드별 유효성 검사(이름, 거리, 설명, 링크 등 입력값 검증),
그리고 추후 추가될 음식점 수정, 삭제, 필터링 등의 기능이 도메인 로직에 해당할 수 있어요.
이런 로직들은 비즈니스 규칙을 포함하고 있기 때문에 UI와 분리하는 게 좋아요.

이전 미션에서도 경험했듯이, UI(View) 계층은 언제든지 변경될 수 있어요.
예를 들어, 지금은 웹 기반 UI를 사용하고 있지만, 나중에 CLI 환경으로 바뀔 수도 있고, 디자인 요구사항이 변경되면서 UI 구조가 달라질 수도 있어요.
이런 UI 변경이 생길 때 도메인 로직이 영향을 받지 않도록 독립적으로 관리하는 게 중요해요.

이런 점을 고려하면, RestaurantService 같은 도메인 모듈을 따로 분리하는 게 좋아요.
이 모듈이 음식점 목록을 관리하면서, 추가, 수정, 삭제, 유효성 검사 등의 기능을 담당하도록 설계하면 좋아요.
이렇게 하면 UI가 변경되더라도 비즈니스 로직이 영향을 받지 않아서 유지보수성과 확장성이 더 좋아질 거예요.

eunsoA replied:
넵! 말씀해주신 내용 감안해서 현재 RestaurantService로 도메인 로직을 분리했는데요, 혹시 지금보다 더 분리할 도메인 로직이 있을까요?

wonsss replied:
RestaurantService 관련해서는 다른 코멘트에서 말씀드린 부분이 있어서 참고해주세요. 😄 
https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1989749110

wonsss reviewed (changes requested):
안녕하세요. 클레어 😄  이번 미션 리뷰를 맡은 마르코입니다! 🙇 
재사용 가능한 컴포넌트 분리를 잘 고민해주신 것 같아요.
남겨드린 리뷰 참고하시어 개선하시는 데 조금 더 도움이 되셨으면 좋겠네요!
본문에서 질문 주신 내용들은 관련된 맥락의 코드에 코멘트로 달아서 답변드렸으니 참고해주세요.

eunsoA committed:
refactor: 하드코딩된 이미지 경로 imagePaths 파일로 이동

eunsoA committed:
refactor(Header): innerHTML에서 createElement로 변경

- XSS 취약점 방지를 위해 innerHTML 제거
- DOM 조작 방식을 createElement와 appendChild로 변경
- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986038631)

eunsoA committed:
refactor(Button): innerHTML에서 createElement로 변경

- XSS 취약점 방지를 위해 innerHTML 제거
- DOM 조작 방식을 createElement와 appendChild로 변경
- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986038631)
- createButton 오타 수정

eunsoA committed:
refactor(Modal): innerHTML에서 createElement로 변경

- XSS 취약점 방지를 위해 innerHTML 제거
- DOM 조작 방식을 createElement와 appendChild로 변경
- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986038631)

- modal close시, reset 추가
- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986036338)

- createButton 오타 수정
- 모호한 네이밍 수정

eunsoA committed:
style: 모호한 변수 네이밍 수정

- testData 에서 modalFormData로 수정

eunsoA committed:
refactor: querySelector를 호출에서 클래스 필드 저장으로 변경

- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986052669)

eunsoA committed:
refactor: handleBackdropClick 이벤트 위임 적용

eunsoA committed:
refactor(RestaurantIcon): innerHTML에서 createElement로 변경

- XSS 취약점 방지를 위해 innerHTML 제거
- DOM 조작 방식을 createElement와 appendChild로 변경
- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986038631)

eunsoA committed:
refactor(RestaurantInfo): innerHTML에서 createElement로 변경

- XSS 취약점 방지를 위해 innerHTML 제거
- DOM 조작 방식을 createElement와 appendChild로 변경
- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986038631)

eunsoA committed:
fix: 변수명 변경에 따른 미반영 수정

- testData -> modalFormData 수정

eunsoA committed:
fix: 추가하기 클릭이벤트 로직 수정

eunsoA committed:
refactor: 모달창 esc 클릭 이벤트 위임

- AddRestaurantModal에서 Modal로 Escape keyup 이벤트 위임

eunsoA committed:
refactor: 사용하지 않는 파일 삭제

- convertStringToElement

eunsoA committed:
refactor: style 폴더 이동

- src/style 에서 src/UI/style/로 이동

eunsoA committed:
refactor: Domain 과 UI로직 분리에 따른 파일 분리

eunsoA committed:
test: Domain 과 UI로직 파일 이동에 따른 수정

eunsoA committed:
refactor(AddRestaurantModal): Domain로직과 UI로직 분리

eunsoA committed:
style: restaurantService.js에서 RestaurantService.js로 변경

eunsoA committed:
fix: 파일명 변경에 따른 Import 문 수정

eunsoA committed:
refactor : Header에 addRestaurantModal 주입 제거

- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986047739)

eunsoA committed:
refactor(Header): header 내부 title과 button 분리

- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986048150)

eunsoA committed:
refactor(Button): button 컴포넌트 type 주입으로 수정

- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986050792)

eunsoA committed:
refactor(Button): Button 컴포넌트의 이벤트 처리 방식 개선

- Button 생성 시 이벤트 리스너를 직접 주입하도록 변경
- 이벤트 바인딩 로직을 private 메서드로 분리
- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986051195)

eunsoA committed:
refactor: CSS 파일 컴포넌트 단위로 분리

- 기존 style.css 파일의 스타일을 각 컴포넌트 폴더로 이동
- 공통 스타일은 style.css로 분리
- 각 컴포넌트에서 CSS 모듈 import 추가
- 리뷰어 리뷰 반영 (https://github.com/woowacourse/javascript-lunch/pull/209#discussion_r1986052205)

eunsoA committed:
fix: 모달의 취소 버튼 이벤트 핸들러 수정

eunsoA committed:
test: 음식점 추가 기능 사용 e2e 테스트 추가

eunsoA commented:
안녕하세요 마르코!
엄청나게 많은 피드백 감사합니닷..!! 
덕분에 많이 생각해보고 리팩터링 해볼 수 있었어요.

위의 댓글에도 몇가지 질문을 남겼지만, 아래에 추가적으로 궁금한 내용이 있어서 질문을 남깁니다!🙋‍♂️

1. AddRestaurantModal의 위치에 대한 고민.
- 저는 2회 이상 재사용을하는 컴포넌트들만 components/에 모았어요. addRestaurantModal은 modal을 상속받기에 modal과 같이 둘 수도 있지만, '2회 이상 사용된다'라는 재사용성에는 부합하지 않는다고 느껴져 modal/ 아래 만들었는데요, 마르코가 보기에는 구조상 이 위치에 있는게 적합한가요? 아니면 다른 방법이 있을까요?


추가록 혹여나 제가 건강한 리뷰를 위해 놓치고 있는게 있거나 알아야할게 있다면..! 꼭 알려주세요! 
리뷰 감사합니다 :) 🙇

eunsoA committed:
refactor: css파일 html 직접 로드로 변경

eunsoA committed:
fix: 모달창 추가하기 버튼 이벤트 수정

- 필수항목 미입력시 alert후 모달창 닫히는 문제 해결

wonsss commented:
> AddRestaurantModal의 위치에 대한 고민.
저는 2회 이상 재사용을하는 컴포넌트들만 components/에 모았어요. addRestaurantModal은 modal을 상속받기에 modal과 같이 둘 수도 있지만, '2회 이상 사용된다'라는 재사용성에는 부합하지 않는다고 느껴져 modal/ 아래 만들었는데요, 마르코가 보기에는 구조상 이 위치에 있는게 적합한가요? 아니면 다른 방법이 있을까요?

좋은 고민이네요. 이것도 정답은 없지만, 논리적으로 이러한 고민을 하고 방향성을 스스로 잡아나가면 좋은 것 같아요.
고민하신 것처럼 재사용 가능한 컴포넌트와 특정 페이지에서만 사용되는 일회성 컴포넌트를 분리해볼 수 있죠. 이와 같이하면 유지보수성과 가독성이 좋아지고, 컴포넌트의 역할이 명확해져요. 
말씀대로 최상위 레벨의 components/ 는 여러 페이지에서 재사용되는 컴포넌트를 두는 폴더로 둬도 좋을 것 같아요. 
특정 페이지에서만 사용되는 컴포넌트는 pages/lunch/components/ 처럼 특정 페이지의 폴더 하위에 components 폴더를 두는 방법이 있어요.

wonsss reviewed (approved):
안녕하세요. 클레어! 남겨드린 리뷰를 다 꼼꼼히 읽어봐주시고 최선을 다해 반영해주신 것 같아 기쁩니다. 👍 
반영해주신 내용 살펴봤고, 추가로 피드백 드릴 부분에는 몇 가지 코멘트 달아드렸어요. 추가로 드린 코멘트 중 혹시 반영하고 싶으신 부분은 2단계에서 수정해주셔도 될 것 같아요.
이미 충분히 1단계 목표 및 리뷰 의도를 잘 이루셔서 여기서 1단계 마무리할게요.  수고하셨습니다! 😄

eunsoA commented:
열심히 리뷰 내용에 대해 고민하고 반영해봤는데 잘 봐주셔서 감사합니다 ☺️ 

추가적으로 말씀해주신 부분은 보완해서 2단계 미션에서 뵐게요!
좋은하루 보내세요 ! :)
